File: digital_clock.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <signal.h>
#include "PCD8544.h"
#include "wiringPi/wiringPi/wiringPi.h"

uint8_t pin_sclk = 4;
uint8_t pin_din = 3;
uint8_t pin_dc = 2;
uint8_t pin_rst = 0;
uint8_t pin_ce = 1;

// This is already tuned
uint8_t lcd_contrast = 60;
char timeString[9];

void gpio_cleanup(void)
{
    pinMode(pin_sclk, INPUT);
    pinMode(pin_din, INPUT);
    pinMode(pin_dc, INPUT);
    pinMode(pin_rst, INPUT);
    pinMode(pin_ce, INPUT);
}

char * get_time(void)
{
    time_t current_time;
    struct tm * time_info;
    time(&current_time);
    time_info = localtime(&current_time);
    strftime(timeString, sizeof(timeString), "%H:%M:%S", time_info);
    // printf("%s\n", &timeString);
    return (char *) &timeString;
}

void signal_callback(int signum)
{
    printf("Caught signal: %i\n", signum);
    LCDclear();
    LCDdisplay();
    gpio_cleanup();
    exit(0);
}

int main(int argc, char const *argv[])
{
    printf("nLCD tool\n");
    printf("Pin definations:\n");
    printf("CLK on %i\n", pin_sclk);
    printf("DIN on %i\n", pin_din);
    printf("DC on %i\n", pin_dc);
    printf("CE on %i\n", pin_ce);
    printf("RST on %i\n", pin_rst);

    if (wiringPiSetup() == -1)
    {
        printf("wiringPi error\n");
        exit(1);
    }
    signal(SIGINT, signal_callback);

    LCDInit(pin_sclk, pin_din, pin_dc, pin_ce, pin_rst, lcd_contrast);
    LCDcommand(PCD8544_DISPLAYCONTROL | PCD8544_DISPLAYALLON);
    LCDdisplay();
    LCDclear();
    LCDcommand(PCD8544_DISPLAYCONTROL | PCD8544_DISPLAYNORMAL);

    while (1)
    {
        LCDdrawrect(6 - 1, 6 - 1, LCDWIDTH - 6, LCDHEIGHT - 6, BLACK);
        LCDdrawstring(20, 12, get_time());
        LCDdisplay();
        delay(1000);
    }
}

File: digital_clock.go
package main

import (
	"fmt"
	"os"
	"os/signal"
	"strconv"
	"syscall"
	"time"

	"github.com/JokerQyou/rpi"
	"github.com/JokerQyou/rpi/pcd8544"
	"github.com/kidoman/embd"
	"github.com/kidoman/embd/sensor/bmp085"

	_ "github.com/kidoman/embd/host/rpi"
)

const (
	SCLK     = 0
	DIN      = 1
	DC       = 2
	CS       = 3
	RST      = 4
	CONTRAST = 60
)

func init() {
	rpi.WiringPiSetup()
	pcd8544.LCDInit(SCLK, DIN, DC, CS, RST, CONTRAST)
	pcd8544.LCDclear()
	pcd8544.LCDdisplay()

}

func gpio_cleanup() {
	pcd8544.LCDclear()
	pcd8544.LCDdisplay()

	rpi.PinMode(SCLK, rpi.INPUT)
	rpi.PinMode(DIN, rpi.INPUT)
	rpi.PinMode(DC, rpi.INPUT)
	rpi.PinMode(CS, rpi.INPUT)
	rpi.PinMode(RST, rpi.INPUT)
}

func get_time() (string, string) {
	t := time.Now()
	return t.Format("15:04:05"), t.Format("01-02 Mon")
}

func main() {
	sig := make(chan os.Signal, 1)
	if err := embd.InitI2C(); err != nil {
		panic(err)
	}
	defer embd.CloseI2C()

	bus := embd.NewI2CBus(1)
	baro := bmp085.New(bus)
	defer baro.Close()

	// Draw the outline
	pcd8544.LCDdrawrect(6-1, 6-1, pcd8544.LCDWIDTH-6, pcd8544.LCDHEIGHT-6, pcd8544.BLACK)
	// date string changes slowly
	var (
		date_str string
		temp_str string
	)

	signal.Notify(sig, syscall.SIGINT, syscall.SIGTERM)
	go func() {
		s := <-sig
		fmt.Println(s)
		gpio_cleanup()
		os.Exit(0)
	}()

	keep_running := true
	for keep_running {
		// Get temperature
		temp, err := baro.Temperature()
		if err != nil {
			gpio_cleanup()
			panic(err)
		}
		_temp_str := fmt.Sprint(strconv.FormatFloat(temp, 'f', 2, 64), " C")

		time_str, _date_str := get_time()
		pcd8544.LCDdrawstring(20, 12, time_str)
		if date_str != _date_str {
			date_str = _date_str
			pcd8544.LCDdrawstring(18, 24, date_str)
		}
		if temp_str != _temp_str {
			temp_str = _temp_str
			pcd8544.LCDdrawstring(20, 36, temp_str)
		}
		pcd8544.LCDdisplay()
		// wait for 1 sec
		time.Sleep(time.Second)
	}
}
